COMPUTE-1553 add allowOverwrite readme (#107)
dnanexus/dxfuse: README.md

`dxfuse -limitedWrite` mode was primarily designed to support spark file output over the `file:///` protocol.

Creating then writing to new files is allowed when dxfuse is mounted with the `-limitedWrite` flag.
Writing to these newly created files is **append only**. Any non-sequential writes will return `ENOTSUP`. Seeking or reading operations are not permitted while a file is being written.

## Supported operations

> `-limitedWrite` mode enables the following operations: rename (mv, see [below](#rename-behavior)), unlink (rm), mkdir (see [below](#mkdir-behavior)), and rmdir (empty folders only). Rewriting existing files is not permitted, nor is truncating existing files, unless `-allowOverwrite` is also set. See [below](#allowoverwrite-mode).

Reviewer: ```suggestion
`-limitedWrite` mode enables the following operations: rename (mv, see [below](#rename-behavior)), unlink (rm), mkdir (see [below](#mkdir-behavior)), and rmdir (empty folders only). Rewriting existing files is not permitted, nor is truncating existing files, unless `-allowOverwrite` is also specified. See [below](#allowoverwrite-mode).
```

|	37.8  |	87 | 10GiB |
|	254  |	495 | 100GiB |

> # AllowOverwrite Mode

Reviewer: ```suggestion
# `-allowOverwrite` Mode
```

> If `-limitedWrite` and `-allowOverwrite` flags are both set when launching dxfuse, overwriting existing files (both created before or during the current session) is enabled, but **only in truncate mode** (with `O_TRUNC` flag at system call level)

Reviewer: ```suggestion
`-allowOverwrite` flag requires `-limitedWrite` flag to be specified as well.

Launching dxfuse with `-limitedWrite -allowOverwrite` flags allows dxfuse clients to truncate and overwrite existing files created both before and during the current dxfuse session (files must be opened with`O_TRUNC` flag at system call level) in addition to other operations allowed by `-limitedWrite`.  Append-only limitations of the `-limitedWrite` mode apply to the overwritten files.
```

Sample use cases:

> - In terminal, write to an existing file by redirecting using '>' or 'tee'

Reviewer: ```suggestion
- In a terminal, overwrite an existing file using `>`, `tee`, `cp`, `dd`
```

> - In terminal, overwrite an existing file by `cp` or `dd`

Reviewer: ```suggestion
```

> - write and save a file in RStudio editor

Reviewer: ```suggestion
- Save a modified file in RStudio editor
```

> - overwrite an existing file using R commands from an R script or in R console, for example:

Reviewer: ```suggestion
- Overwrite an existing file using the following R commands:
```

  - base::writeLines()
  - data.table::fwrite()
  - readr::write_rds()
  - saveRDS()
  - write.table()

> Other operations enabled by `-limitedWrite` mode are also supported. However, any other operations that modifies an existing file content without opening in truncate mode is **not supported**, for example:

Reviewer: ```suggestion
Operations that attempt to modify an existing file without opening it in truncate mode are not supported. For example the following operations operations will fail:
```

> - write to an existing file by redirecting using '>>' in terminal

Reviewer: ```suggestion
- In a terminal, appending to an existing with `>>`
```

> - open existing file with `O_APPEND` flag

Reviewer: ```suggestion
- Opening an existing file with `O_APPEND` flag
```

Reviewer: @YuxinShi0423 what if O_TRUNC and O_APPEND are both specified?  Is that possible?

> - truncate file to size 0 using `truncate` command or other commands that call `ftruncate` method

Reviewer: ```suggestion
- Truncating and existing file to size 0 using `truncate` command or other commands that call `ftruncate` system call
```

> Note: if using `vim` to edit existing file, it is suggested to avoid swap file creation by doing either of the following:

Reviewer: ```suggestion
Note: When using `vim` to edit existing dxfuse-backed files, avoid swap file creation in dxfuse-backed folders by doing either of the following:
```

Reviewer: @YuxinShi0423  
What's the drawback of  not following these instructions?  

Does vim editing still work, just in an inefficient way?

Author: It would show errors like "E297: Write error in swap file", or warning a lingering swap file when reopening the file. The editing still works but error-prone and inefficient.

> - turn off swap file: put `set noswapfile` in `~/.vimrc` 

Reviewer: ```suggestion
- turn off swap file by adding `set noswapfile` to your `~/.vimrc` 
```

> - set the directory option to the location outside the mounted folder

Reviewer: ```suggestion
- set the directory option to a non-dxfuse folder
```

  ```

>   mkdir -p ~/.vim/swapfiles  # create a temp dir to store swap files

Reviewer: ```suggestion
  mkdir -p ~/.vim/swapfiles  # create a temp dir to store swap files outside of the dxfuse folder
```

>   echo "set directory=$HOME/.vim/swapfiles" > ~/.vimrc

Reviewer: ```suggestion
  echo "set directory=$HOME/.vim/swapfiles" >> ~/.vimrc
```
